**Group 08**

Ana Cláudia Alferes - M20190932


Francisco Carujo Neves - M20190089


Luís Filipe Pinho - M20190935

# Big Data Modeling and Management Assigment


## Secong Assigment

For this first project we will be taking a european public procurement notices database to analyse contracts and money expenditure within the European Union!   

For reference the dataset used for this database has been extracted from [Data Europa](https://data.europa.eu/euodp/en/data/dataset/ted-csv), and the definition of this data can be found [here](https://data.europa.eu/euodp/repository/ec/dg-grow/mapps/2019/TED(csv)_data_information_v3.2.doc). 

#### Problem description

Explore the database via python's mongoDB connector and/or the ROBO3T software.   
Answer the questions.  
Submit the results by following the instructions

#### Connection details to the MongoDB database
Each group should have received by email the credentials to connect to the group's mongo database.
```
Connection example: mongodb://username:password@host:port
```
The database contains 4 collections.


### Questions

#### To use the find Method:
1. How many contracts are in the database?
1. What is the ('AWARD_VALUE_EURO') of the highest value contract ('AWARD_VALUE_EURO') in the dataset?
1. In which years were the first and last contracts celebrated?
1. What was the 'CRIT_CRITERIA' of the highest value contract ('AWARD_VALUE_EURO') in Portugal in 2010?
1. What was the 'CRIT_CRITERIA' of the highest number of offers ('NUMBER_OFFERS') using European funds?
1. Number of contracts that have 'Defence' in the 'MAIN_ACTIVITY' value string?
1. Number of contracts where both companies are from the same town?
1. Number of contracts where difference between Proposed value ('VALUE_EURO') and awarded value ('AWARD_VALUE_EURO') is greater than 10000000? 

#### To use the Aggregate Framework:
9. Which country has the most contracts overall?
1. Which year had the highest money transactions value ('AWARD_VALUE_EURO')?

##### Knowing that the cpv division is represented in the first 2 digits of the CPV code ('CPV'):
11. Group by year, country and CPV division (this order) and return the average/max and counts of contract award values

##### Open Question
12. Knowing that the dataset has wrongly inserted values, more in particular impossible 'VALUE_EURO' and similar values, how would you deal with this situation?



### Group  

This project assumes groups to be the same as the previous project, any copying detected by the professors will lead to a grading of zero on the project and/or other disciplinary actions!


### Submission      

Each group will have its own database to work with. Use CRUD operations to **create a new collection, called 'results' where you will input your answers**  with the following structure:  

- Each answer should be inserted as a document, all documents should have a key 'question' (no caps) were its value is the questions number, and another key called 'answer' where naturally the value should be the answer to that particular question;  

- Question 11 output is a list which can also be inserted in the answer value;  

- Question 12 should follow the structure of the 'single answer' questions (we expect here a text explaining your strategies to deal with this issue) 


Example:
```
db["results"].insert_one(
    {'question' : 0,
     'answer' : result
    },
)
```

To evaluate results:

```
list(db["results"].find( {'answer' : 0}))

Output: [{'_id': ObjectId('5eb53aa694c09a2299fb069e'), 'question': 0, 'answer': result}]
```

Delivery date: Until **23:59 of May 21st**


### Evaluation    


This will be 15% of the final grade.   
Each solution will be evaluate on 2 components, each equally important: correctness of results and simplicity of the solution.  
All code will go through plagiarism automated checks. Groups with the same code will undergo investigation

In [1]:
from pymongo import MongoClient

In [2]:
host='rhea.isegi.unl.pt'
port='28008'
user='GROUP_8'
password= 'MjY4NDI2MDIwMzE5MjU5NjczNzE5ODMxNTk4MDkxMDAxMDEzMTAx'
protocol="mongodb"
client = MongoClient(f"{protocol}://{user}:{password}@{host}:{port}")

In [3]:
db = client.contracts

### QUESTION 1

In [4]:
result_1 = db.eu.find().count()
result_1

D:\Programs\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  """Entry point for launching an IPython kernel.


5563372

To answer this question, we used the ‘eu’ collection. Since every row is a different contract, we called the collection, and with the method ‘find’, we counted the number of rows in ‘eu’, giving the total number of contracts. So, the answer to this question is a total of 5563372 contracts in the database.

### QUESTION 2

In [5]:
result_2 = list(db.eu.find(sort=[('AWARD_VALUE_EURO',-1)]).limit(1))[0]['AWARD_VALUE_EURO']
result_2

1e+20

Using the collection ‘eu’, which has information on every contract, we sorted the values of ‘AWARD_VALUE_EURO’ in descending order, and limited the output just for the first row, which was the contract with the highest award. And finally, for the output, we only returned the value from the requested column. The contract with the highest AWARD VALUE EURO’, has a value of 1e+20€.

### QUESTION 3

In [6]:
result_3_1 = list(db.eu.find(sort=[('YEAR',1)]).limit(1))[0]['YEAR']
result_3_1

2009

In [7]:
result_3_2 = list(db.eu.find(sort=[('YEAR',-1)]).limit(1))[0]['YEAR']
result_3_2

2018

For this question, we formulated two queries. The first query returns the last contract celebrated in this database. Using always the collection ‘eu’, and the method ‘find’, we sorted the contracts in descending order by the column ‘YEAR’, and limited the output to the first row of the column ‘YEAR’, which is the highest value. So, the last contract celebrated was in 2018. As for the first contract, the process was similar, but instead of sorting the values in descending order, we sorted in ascending order, so, the first row would return the smallest value for the variable ‘YEAR’. The first contract celebrated in this database was in 2009.

### QUESTION 4

In [8]:
result_4 =db.eu.find({'ISO_COUNTRY_CODE':'PT','YEAR':2016},sort=[('AWARD_VALUE_EURO',-1)]).limit(1)[0]['CRIT_CRITERIA']
result_4

'Preço por quilómetro percorrido entre o Ponto B e o Ponto C---Percentagem do preço a aplicar à parte do percurso entre o Ponto A — local de saída do prestador — e o Ponto B — local onde se inicia o transporte do doente –, nas deslocações superiores a 20 km---Preço por viagem menor ou igual a 20 km por doente e/ou acompanhante que inclui deslocações de ida e volta («taxa de saída»)---Percentagem do preço a pagar a partir do segundo doente para deslocações superiores a 20 km e inferiores a 100 km, inclusive---Percentagem do preço a pagar a partir do segundo doente para deslocações superiores a 100 km---Percentagem do preço a pagar por cada acompanhante, médico/enfermeiro---Preço máximo a pagar pela segunda hora de espera e subsequentes---Preço do kit de parto---Preço do ventilador---Preço do oxigénio e Capacidade de transporte do concorrente'

In this question, we had some constraints that have to be considered. Using the collection ‘eu’ and the method ‘find’, we constrained the database for the ISO_COUNTRY_CODE’ to ‘PT’ (Portugal), and the ‘YEAR’ to 2016. For the ‘AWARD_VALUE_EURO’, we sorted in descending order, and limited to the first row, which corresponds to the contract with highest award. Finally, since it was only requested the ‘CRIT_CRITERIA’, we returned only the value for that column.

### QUESTION 5

In [9]:
result_5 = list(db.eu.find({'B_EU_FUNDS':'Y'},sort=[('NUMBER_OFFERS',-1)]).limit(20))[0]['CRIT_CRITERIA']
result_5

'Cost---Methodology & resources---Understanding of the sensitivities of working on a designated site---Understanding of the difficulties of working on islands'

In this question we started by restricting our demand on contracts with European funds only, then we sorted by the number of offers to find the highest one. After finding the contract with European funds and the highest number of offers we returned the "CRIT_CRITERIA" field. 

### QUESTION 6

In [10]:
result_6 = db.eu.find({'MAIN_ACTIVITY':"Defence"}).count()
result_6

D:\Programs\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  """Entry point for launching an IPython kernel.


120098

In this code the objective was to find the number of contracts that had "Defence" in their "MAIN_ACTIVITY" field and we did that by using a count in the condition required inside the find.

### QUESTION 7

In [11]:
result_7 = db.eu.find({"$expr":{"$eq":["$WIN_TOWN", "$CAE_TOWN"]}}).count()
result_7

D:\Programs\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  """Entry point for launching an IPython kernel.


721751

We can see in question that we are using a expression to find the contracts were both companies are from the same town. We finally check the exact number of times this happens if the dataset using a count operator.

### QUESTION 8

In [12]:
result_8 = db.eu.find({ '$expr': {'$gt': [ {"$subtract":["$VALUE_EURO", "$AWARD_VALUE_EURO"]} , 10000000] }}).count()
result_8

D:\Programs\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: count is deprecated. Use Collection.count_documents instead.
  """Entry point for launching an IPython kernel.


110928

In this question we are supposed to arrive at the total number of contracts were the difference between Proposed value (‘VALUE_EURO’) and awarded value (‘AWARD_VALUE_EURO’) is greater than 10000000. So we start by using a expression were we want the value inside the "gt" (greater than) expression , that corresponds to the subtraction of the "VALUES_EURO" with the "AWARD_VALUE_EURO" fields, to be greater than 10000000. Finally to get to our final answer we made the count operator.

### QUESTION 9

In [13]:
result_9 = list(db.eu.aggregate([
    {"$group" : {'_id':"$ISO_COUNTRY_CODE", 'count':{'$sum':1}}}
    ,{ "$sort": {"count": -1 }},{"$limit": 1}
]))
result_9

[{'_id': 'FR', 'count': 1263013}]

To answer this and the following two questions, we make use of aggregate functions.
For question 9, we first group it by the value of ISO_COUNTRY_CODE, as this will represent a country’s designation, and by its count. On the latter part we are essentially saying each row represents a 1 and adding all those values. After this is done, we get an output which has the count of all countries’ contracts. To check which one has the greatest number of contracts we sort it by -1, which means that we sort it in descending, and limit the values to only 1 row. This outputs the country alongside the number of contracts.


### QUESTION 10

In [14]:
result_10 = list(db.eu.aggregate([
    {"$group" : {'_id':"$YEAR", 'total':{'$max':"$AWARD_VALUE_EURO"}}}
    ,{ "$sort": {"total": -1 }},{"$limit": 1}
]))
result_10

[{'_id': 2012, 'total': 1e+20}]

For question 10 the process is quite like previous question. We this time group it by year and its maximum value. As we want the highest value per year, to then compare all other years, it only makes sense we use max, as this will output the max value per year of AWARD_VALUE_EURO. Then, like in the last question we sort it by the value of this new value outputted and limit it only to the highest one.

### QUESTION 11

In [15]:
result_11 = list(db.eu.aggregate([
       {
     '$unwind': { 'path': "$AWARD_VALUE_EURO"}
   },
    { "$group": {"_id": {"year": "$YEAR",'country':"$ISO_COUNTRY_CODE","CPV": {'$substr':[{'$toString':'$CPV'},0,2]}},"avg" : {"$avg":"$AWARD_VALUE_EURO"}, "max" : {"$max":'$AWARD_VALUE_EURO'},"count": {"$sum": 1 }
    }},{ "$sort": {"count": -1 }}
]))

In [16]:
result_11

[{'_id': {'year': 2017, 'country': 'PL', 'CPV': '33'},
  'avg': 196830.7314406318,
  'max': 1242592592.6,
  'count': 84456},
 {'_id': {'year': 2018, 'country': 'PL', 'CPV': '33'},
  'avg': 60144.620385805676,
  'max': 235807637.71,
  'count': 81751},
 {'_id': {'year': 2015, 'country': 'PL', 'CPV': '33'},
  'avg': 37182.45358646133,
  'max': 22155788.25,
  'count': 80776},
 {'_id': {'year': 2014, 'country': 'PL', 'CPV': '33'},
  'avg': 36560.12772191674,
  'max': 30948064.81,
  'count': 77653},
 {'_id': {'year': 2013, 'country': 'PL', 'CPV': '33'},
  'avg': 35081.32442354463,
  'max': 33112555.57,
  'count': 77300},
 {'_id': {'year': 2016, 'country': 'PL', 'CPV': '33'},
  'avg': 42050.18444680089,
  'max': 22944002.29,
  'count': 72614},
 {'_id': {'year': 2012, 'country': 'PL', 'CPV': '33'},
  'avg': 44160.95515101423,
  'max': 354658032.83,
  'count': 64729},
 {'_id': {'year': 2011, 'country': 'PL', 'CPV': '33'},
  'avg': 44276.739632628654,
  'max': 79331196.43,
  'count': 61872},
 {'

Finally, in response to question 11, we first unwind AWARD_VALUE_EURO. This ensures that we will not be dealing with any rows that do not have that value, since on default unwind outputs all the rows that have the value present in path. After doing this we apply our criteria. We once again use a group to get the variables year, country and cpv code, assigning these 3 to be the id for the question, alongside the average and max of the contracts value, and the contracts count. To get the CPV code we use a substring function as we know that this code is present on the first two characters of the value, though we first have to convert it into a string, as the usage of the substring function is not possible with a number. After that we simply apply the avg and max functions to AWARD_VALUE_EURO and a count, which is explained in question 9. Finally, we just sort this by count so we can have some order to our list, although this last part is not necessary.

### QUESTION 12

In [17]:
result_12='Throughout answering the various questions posed in this homework we came to realize that there existed some problems with the database. One of these problems was the values of some variables, namely in AWARD_VALUE_ERROR. With this variable we found values such as 1e+20, a value too high to write in its totality. These impossible values end up posing some problems to our analysis, as these probably refer to the wrongly inserted values mentioned in the question. As such the best way to solve this problem would be to check for outliers in numeric variables. By doing this we can find where these impossible values lie and eliminate certain rows above a certain threshold that we deem acceptable following the outlier analysis. By doing this we end up removing these exceptions and can then proceed to a better analysis. We can also use common sense to remove some impossible values such as possible negative values in contracts. Another problem we verified was found in question 4 before it was rectified. There are some rows that don’t have all the variables, as was the case in question 4, where before it was corrected, the highest value of AWARD_VALUE_EURO did not contain the variable CRIT_CRITERIA, which deemed the question unanswerable. This problem was encountered again in question 11, with several AWARD_VALUE_EURO variables not found. This meant that when doing a count we had all the contracts counted, even the ones without this variable, but the functions max and avg did not take these values into account, which ended up tampering with our analysis initially. The quickest answer we found for this problem was just putting these rows without the required variable aside, with the possibility of eliminating them. However, if the number of rows eliminated doing this is too great it ends up affecting other results. As such another possible strategy would be to fill these values, with averages if it is a numeric value, or a standard placeholder text for strings.'

Throughout answering the various questions posed in this homework we came to realize that there existed some problems with the database. One of these problems was the values of some variables, namely in AWARD_VALUE_ERROR. With this variable we found values such as 1e+20, a value too high to write in its totality. These impossible values end up posing some problems to our analysis, as these probably refer to the wrongly inserted values mentioned in the question. As such the best way to solve this problem would be to check for outliers in numeric variables. By doing this we can find where these impossible values lie and eliminate certain rows above a certain threshold that we deem acceptable following the outlier analysis. By doing this we end up removing these exceptions and can then proceed to a better analysis. We can also use common sense to remove some impossible values such as possible negative values in contracts. Another problem we verified was found in question 4 before it was rectified. There are some rows that don’t have all the variables, as was the case in question 4, where before it was corrected, the highest value of AWARD_VALUE_EURO did not contain the variable CRIT_CRITERIA, which deemed the question unanswerable. This problem was encountered again in question 11, with several AWARD_VALUE_EURO variables not found. This meant that when doing a count we had all the contracts counted, even the ones without this variable, but the functions max and avg did not take these values into account, which ended up tampering with our analysis initially. The quickest answer we found for this problem was just putting these rows without the required variable aside, with the possibility of eliminating them. However, if the number of rows eliminated doing this is too great it ends up affecting other results. As such another possible strategy would be to fill these values, with averages if it is a numeric value, or a standard placeholder text for strings.

# SUBMISSION 

In [18]:
db["results"].insert_one(
    {'question' : 1,
     'answer' : result_1
    },
)

In [19]:
db["results"].insert_one(
    {'question' : 2,
     'answer' : result_2
    },
)

In [20]:
db["results"].insert_one(
    {'question' : 3,
     'answer' : [result_3_1,result_3_2]
    },
)

In [21]:
db["results"].insert_one(
    {'question' : 4,
     'answer' : result_4
    },
)

In [22]:
db["results"].insert_one(
    {'question' : 5,
     'answer' : result_5
    },
)

In [23]:
db["results"].insert_one(
    {'question' : 6,
     'answer' : result_6
    },
)

In [24]:
db["results"].insert_one(
    {'question' : 7,
     'answer' : result_7
    },
)

In [25]:
db["results"].insert_one(
    {'question' : 8,
     'answer' : result_8
    },
)

In [26]:
db["results"].insert_one(
    {'question' : 9,
     'answer' : result_9
    },
)

In [27]:
db["results"].insert_one(
    {'question' : 10,
     'answer' : result_10
    },
)

In [28]:
db["results"].insert_one(
    {'question' : 11,
     'answer' : result_11
    },
)

In [29]:
db["results"].insert_one(
    {'question' : 12,
     'answer' : result_12
    },
)